docs/alerts.md

```diff
@@ -0,0 +1,312 @@
+on receiving the next point after 10 seconds.
+
+If we were in a situation where the handler were to skip an action because a future notification pre-empted it (such as if only a single table were received and the handler noticed that the alert was already resolved), the `_at` time would be null since the action never occurred, but the row would still be present.
+
+#### Additional Notes
+
+As can be seen, the `notify()` method is simply a placeholder to allow notification handlers to be abstracted from the notification itself.
+This lends itself better to code generators because the `notify()` method can be customized using extern properties so that the act of checking and notifying can be separated from eachother.
+
+## Integration Methods
+
+The above constructs lend themselves well to various different methods of integration and also allow for building out future integration plans.
+
+### Inline
+
+It is possible for the entire query to run inline.
+
+```
+// Send to multiple notification handlers.
+notify = (tables=<-) => {
+	tables |> pagerduty.notify(routingKey: myRoutingKey)
+	tables |> slack.notify(channel: "#flux")
+}
+
+from(bucket: "telegraf")
+    |> range(start: -5m)
+    |> filter(fn: (r) => r._measurement == "cpu" and r._field == "usage_system")
+    |> check(
+    	crit: (r) => r._value > 90,
+    )
+    |> notify()
+```
+
+### Using Buckets
+
+Using the same script as the above, the `notify` method can be customized to write to a bucket instead and the notifications can be triggered off of the contents of the bucket instead.
+
+```
+// This goes in the check script.
+notify = (tables=<-) => tables |> to(bucket: "system")
+
+// This goes in a separate script.
+alert = (tables=<-, handler) => tables |> handler() |> to(bucket: "notifications")
+
+tables = from(bucket: "system")
+	|> range(start: -1m)
+	|> notify()
+	
+// The function is easy for the user to write and it is easy to
+// inject into a generated script.
+handler1 = (tables=<-) => pagerduty.notify(routingKey: myRoutingKey)
+handler2 = (tables=<-) => slack.notify(channel: "#flux")
+	
+// Send to multiple notification handlers.
+// This is easy to generate.
+tables |> handler1() |> to(bucket: "notifications")
+tables |> handler2() |> to(bucket: "notifications")
+```
+
+### Tasks
+
+Using the above, this can be integrated with tasks fairly easily and meet the above requirements.
+
+This might be the check script.
+
+```
+// Injected.
+option task = {
+	every: 1m,
+}
+
+// Possibly an option like now?
+notify = (tables=<-) => tables |> to(bucket: "system")
+
+// User provided.
+from(bucket: "telegraf")
+    |> range(start: -task.every)
+    |> filter(fn: (r) => r._measurement == "cpu" and r._field == "usage_system")
+    |> check(
+	crit: (r) => r._value > 90,
+    )
+    |> notify()
+```
+
+This might be the notification script.
+
+```
+// Injected.
+option task = {
+	every: 1m,
+}
+
+tables = from(bucket: "system")
```

Comment:
Reading from the system bucket to get a set of check statuses should be a helper method on the package API. This way the reading and writing to the system bucket is abstracted away from the power user.

Reply:
It is intended that this is generated by tasks as mentioned in the comment.

Reply:
The generated task should be able to call a parameterized helper function that encapsulates the basic concepts. The reason for this is we want to enable power users to leverage the same basic constructs of using checks and notifications without having to manage all the details. Specificaly in this case power users should not need to know that check statuses are stored in a system bucket, rather that they are stored somewhere and can be retrieved using the provided API methods. Who knows, maybe we decide that we should store the statuses in a relational database in the future. Abstracting the details of the storage enables us to change that without breaking the API. 

Something like this is what I have in mind:


```
findChecks = ( predicate,interval) =>
    from(bucket : "system") 
        |> range(start: -interval)
        |> filter(fn: predicate)

```

Then the generated task is simply:

```
option task = {
    every: 5m,
}

checks = findChecks(interval: task.every, predicate: <insert predicate function here from notification rule match field>)

checks
    |> // notification rule logic goes here
    |> slack.notifty()
```